src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java

```diff
@@ -136,19 +137,21 @@ private GenericRecord getPostHookEvent(HookContext hookContext, EventStatus stat
             .set("PerfObject", dumpPerfData(hookContext.getPerfLogger()))
             .set("OperationId", hookContext.getOperationId());
 
-    dumpMapReduceCounters(mrTasks)
+    dumpMapReduceCounters()
         .ifPresent(counters -> recordBuilder.set("MapReduceCountersObject", counters));
-    dumpTezCounters(tezTasks)
+    dumpTezCounters(Utilities.getTezTasks(plan.getRootTasks()))
         .ifPresent(counters -> recordBuilder.set("TezCountersObject", counters));
 
     return recordBuilder.build();
   }
 
-  private static Optional<String> dumpMapReduceCounters(List<ExecDriver> mrTasks) {
-    JSONArray counters =
-        new JSONArray(mrTasks.stream().map(ExecDriver::getCounters).collect(Collectors.toList()));
-
-    return counters.length() > 0 ? Optional.of(counters.toString()) : Optional.empty();
+  private static Optional<String> dumpMapReduceCounters() {
+    // TODO: Use org.apache.hadoop.mapreduce.Counters
+    Object[] countersArray = SessionState.get().getMapRedStats().values().stream()
+        .map(MapRedStats::getCounters)
+        .map(Counters::makeEscapedCompactString)
```

Comment:
`makeEscapedCompactString` does not produce valid JSON, so it will be much harder to parse it on the service side (and even more problematic if `makeEscapedCompactString` logic is changed in different versions). I would suggest remapping counters to the JSON objects/arrays, similarly to TezCounters below.

Reply:
Agreed!